packages/remirror__extension-callout/__stories__/callout.stories.tsx

```diff
@@ -2,74 +2,138 @@ import 'remirror/styles/all.css';
+      document.removeEventListener('click', handleClickEmoji);
+    };
+  }, [handleClickEmoji, handleSelectEmoji]);
+
+  return null;
+};
+
+const renderDialogEmoji = (node: ProsemirrorNode) => {
+  const { emoji: prevEmoji } = node.attrs;
+  const emoji = document.createElement('span');
+  emoji.dataset.emojiContainer = '';
+  emoji.textContent = prevEmoji;
+  emoji.style.cursor = 'pointer';
 
   // Prevent ProseMirror from handling the `mousedown` event so that the cursor
   // won't move when users click the emoji.
-  emoji.addEventListener('mousedown', (e) => {
-    e.preventDefault();
-  });
+  emoji.addEventListener('mousedown', (e) => e.preventDefault());
 
   return emoji;
 };
 
-const renderRandomEmoji = (node: ProsemirrorNode, view: EditorView, getPos: () => number) => {
-  const { emoji: emojiCode, type } = node.attrs;
-  const emoji = document.createElement('img');
-  emoji.style.height = '48px';
-  emoji.style.width = '48px';
-  const blobmoji = new Blobmoji({ data: svgmojiData, type: 'individual' });
-  emoji.src = blobmoji.url(emojiCode);
+const RandomEmoji = () => {
+  const { updateCallout } = useCommands();
+  const { view } = useRemirrorContext();
+  const pos = useRef(-1);
 
-  const choiceRandomEmoji = (currEmoji: string): string => {
+  const choiceRandomEmoji = useCallback((currEmoji: string): string => {
     const availableEmojis = ['😭', '😊', '🥰', '😂', '🙄', '😫', '🤔', '😌', '😍', '🤣'];
     const nextEmoji = availableEmojis[Math.floor(Math.random() * availableEmojis.length)];
     return currEmoji === nextEmoji ? choiceRandomEmoji(currEmoji) : nextEmoji!;
-  };
+  }, []);
 
-  emoji.addEventListener('click', (e) => {
-    e.preventDefault();
+  const handleClickEmoji = useCallback(
+    (e: MouseEvent) => {
+      const target = e.target as HTMLImageElement;
+      e.preventDefault();
 
-    const transaction = view.state.tr.setNodeMarkup(getPos(), undefined, {
-      emoji: choiceRandomEmoji(node.attrs.emoji),
-      type,
-    });
-    view.dispatch(transaction);
-  });
+      if (!target.matches('[data-emoji-container]')) {
+        return;
+      }
+
+      /**
+       * Find the document position of the click element.
+       */
+      pos.current = view.posAtDOM(target, 0);
+
+      updateCallout({ emoji: choiceRandomEmoji(target.alt) }, pos.current);
+    },
+    [view, updateCallout, choiceRandomEmoji],
+  );
+
+  useEffect(() => {
+    document.addEventListener('click', handleClickEmoji);
+
+    return () => {
+      document.removeEventListener('click', handleClickEmoji);
+    };
+  }, [handleClickEmoji]);
+
+  return null;
+};
+const renderRandomEmoji = (node: ProsemirrorNode) => {
+  const { emoji: emojiCode } = node.attrs;
+  const emoji = document.createElement('img');
+  emoji.style.height = '48px';
+  emoji.style.width = '48px';
+  emoji.dataset.emojiContainer = '';
+  emoji.style.cursor = 'pointer';
+  emoji.alt = emojiCode;
+
+  const blobmoji = new Blobmoji({ data: svgmojiData, type: 'individual' });
```

Comment:
Worth storing `new Blobmoji` in a ref?

Reply:
Because `renderRandomEmoji` is not a react functional component, it's a funtion called in createNoveView. So is it unable to use ref in it?

Reply:
🤦 sorry